lib/actions/config.coffee

```diff
@@ -299,12 +304,26 @@ exports.generate =
 				  $ balena help config generate
 			'''
 
+		if !options.application and options.deviceType
+			patterns.exitWithExpectedError '''
+				Specifying a different device type is only supported when
+				generating a config for an application:
+
+				* An application, with --app <appname>
+				* A specific device type, with --device-type <deviceTypeSlug>
+
+				See the help page for examples:
+
+				  $ balena help config generate
+			'''
+
 		Promise.try ->
 			if options.device?
 				return balena.models.device.get(options.device)
 			return balena.models.application.get(options.application)
 		.then (resource) ->
-			balena.models.device.getManifestBySlug(resource.device_type)
+			deviceType = options.deviceType || resource.device_type
+			balena.models.device.getManifestBySlug(deviceType)
```

Comment:
Ideally you'd validate that the given device type is of the same arch as the application's. This is actually the definition of mixed device-type apps -- i.e. different device-types but common architecture. Same applies to the other commands touched by this PR

Reply:
Oh yea that totally makes sense.
This should probably also be in place in the `os configure` action as well I guess.
The API doesn't have such checks atm and the UI & SDK try to enforce this rule on their relevant methods.

Will try to add this as part of this PR, to avoid having users in hard to debug situations.